File: new/dart-maor/next-closure_test.dart
// בדיקת-חוזה (רתמת-זהב) · next-closure — מייבאת אך ורק את האטום-שלה (חוק-4).
// שש דוגמאות-החוזה זהות ביט-אחר-ביט למקור-ה-JS new/atoms/next-closure.test.mjs
// (אותם קלטים→פלטים; הערכים הומרו ל-Dart). אם עובר ⇒ Dart≡JS.
// הרצה: dart run --enable-asserts new/dart-maor/next-closure_test.dart  ⇒ exit 0
// אפס import של dart:convert — ה-JSON נבנה ידנית מסדר-ההכנסה (מדמה JSON.stringify).
import 'next-closure.dart';

// — סריאליזציה זהה ל-JSON.stringify עבור הדאטה הזה: null/String/bool/num/List/Map,
//   סדר-הכנסה נשמר, תווי-עברית כלשונם. —
String _stringify(dynamic v) {
  if (v == null) return 'null';
  if (v is String) {
    final esc = v.replaceAll('\\', '\\\\').replaceAll('"', '\\"');
    return '"$esc"';
  }
  if (v is bool) return v ? 'true' : 'false';
  if (v is num) return v.toString();
  if (v is List) return '[${v.map(_stringify).join(',')}]';
  if (v is Map) {
    final parts = <String>[];
    v.forEach((k, val) => parts.add('${_stringify(k as String)}:${_stringify(val)}'));
    return '{${parts.join(',')}}';
  }
  throw StateError('unsupported type: ${v.runtimeType}');
}

void main() {
  // שקעי-דמה מקומיים לבדיקה (כמו במקור-ה-JS).
  final CITIES = <String, dynamic>{
    'jerusalem': {'he': 'ירושלים', 'lat': 31.778, 'lon': 35.235, 'candle': 40},
    'telaviv': {'he': 'תל אביב', 'lat': 32.083, 'lon': 34.8, 'candle': 18},
  };
  final WIN = <String, dynamic>{
    'reason': 'שבת', 'kind': 'shabbat', 'startIso': '2026-08-28',
    'startTime': '18:42', 'endIso': '2026-08-29', 'endTime': '19:53', 'days': 1,
  };

  // 1. בלי telephony ⇒ null והשקע לא נקרא
  {
    var called = 0;
    final r = nextClosure(<String, dynamic>{}, '2026-08-24',
        (a, b, c, d) { called++; return [WIN]; }, CITIES);
    assert(r == null, '1: לא null — ${_stringify(r)}');
    assert(called == 0, '1: השקע נקרא למרות שאין telephony ($called)');
  }

  // 2+6. עיר מוכרת + מיפוי-שדות + אימות ארגומנטים-לשקע
  {
    List<dynamic>? gotArgs;
    final r = nextClosure({'telephony': {'city': 'telaviv'}}, '2026-08-24',
        (a, b, c, d) { gotArgs = [a, b, c, d]; return [WIN]; }, CITIES);
    assert(
        _stringify(r) ==
            '{"reason":"שבת","kind":"shabbat","startIso":"2026-08-28","candle":"18:42","endIso":"2026-08-29","tzeis":"19:53","cityHe":"תל אביב"}',
        '2: פלט לא תואם — ${_stringify(r)}');
    assert(
        _stringify(gotArgs) ==
            '["2026-08-24",10,{"city":"telaviv","timezone":"Asia/Jerusalem"},{}]',
        '6: ארגומנטי-השקע לא תואמים — ${_stringify(gotArgs)}');
  }

  // 3. אין חלונות ⇒ null
  {
    final r = nextClosure({'telephony': {'city': 'telaviv'}}, '2026-08-24',
        (a, b, c, d) => <dynamic>[], CITIES);
    assert(r == null, '3: לא null — ${_stringify(r)}');
  }

  // 4. בלי city ⇒ tenant.city='default' + נפילת-ירושלים
  {
    List<dynamic>? gotArgs4;
    final r4 = nextClosure({'telephony': <String, dynamic>{}}, '2026-08-24',
        (a, b, c, d) { gotArgs4 = [a, b, c, d]; return [WIN]; }, CITIES);
    assert(
        _stringify(gotArgs4![2]) == '{"city":"default","timezone":"Asia/Jerusalem"}',
        '4: tenant ברירת-מחדל לא תואם — ${_stringify(gotArgs4![2])}');
    assert(r4!['cityHe'] == 'ירושלים', '4: cityHe ≠ ירושלים — ${r4!['cityHe']}');
  }

  // 5. עיר לא-מוכרת ⇒ tenant שומר אותה, cityHe נופל לירושלים
  {
    List<dynamic>? gotArgs5;
    final r5 = nextClosure({'telephony': {'city': 'nowhere'}}, '2026-08-24',
        (a, b, c, d) { gotArgs5 = [a, b, c, d]; return [WIN]; }, CITIES);
    assert((gotArgs5![2] as Map)['city'] == 'nowhere',
        '5: tenant.city ≠ nowhere — ${(gotArgs5![2] as Map)['city']}');
    assert(r5!['cityHe'] == 'ירושלים', '5: cityHe ≠ ירושלים — ${r5!['cityHe']}');
  }

  print('✓ next-closure: 6 דוגמאות-חוזה — ירוק (Dart≡JS · 2 שקעים מוזרקים, אפס import פנימי)');
}

File: new/atoms/next-closure.test.mjs
import { nextClosure as __pure_nextClosure } from './next-closure.mjs';
// צילום-מקומי + עטיפת-כריכה (מנוע-הטיהור v3)
const __d_nextClosure_NEXT_CLOSURE_T = {
  k1: "default",
  k2: "Asia/Jerusalem",
};
const nextClosure = (...a) => __pure_nextClosure(...a, ...Array(Math.max(0, 4 - a.length)).fill(undefined), __d_nextClosure_NEXT_CLOSURE_T);
// שקעי-דמה מקומיים לבדיקה — הבדיקה מייבאת רק את האטום שלה (חוק-4).
const CITIES = {
  jerusalem: { he: 'ירושלים', lat: 31.778, lon: 35.235, candle: 40 },
  telaviv: { he: 'תל אביב', lat: 32.083, lon: 34.8, candle: 18 },
};
const WIN = { reason: 'שבת', kind: 'shabbat', startIso: '2026-08-28', startTime: '18:42', endIso: '2026-08-29', endTime: '19:53', days: 1 };
let f = 0;
const eq = (name, got, want) => {
  const g = JSON.stringify(got), w = JSON.stringify(want);
  if (g !== w) { console.error(`✗ ${name}: ${g} ≠ ${w}`); f = 1; }
};

// 1. בלי telephony ⇒ null והשקע לא נקרא
let called = 0;
eq('דוגמה 1 (בלי telephony)', nextClosure({}, '2026-08-24', () => { called++; return [WIN]; }, CITIES), null);
if (called !== 0) { console.error('✗ דוגמה 1: השקע נקרא למרות שאין telephony'); f = 1; }

// 2+6. עיר מוכרת + מיפוי-שדות + אימות ארגומנטים-לשקע
let gotArgs = null;
const hcw2 = (...args) => { gotArgs = args; return [WIN]; };
eq('דוגמה 2 (תל אביב)', nextClosure({ telephony: { city: 'telaviv' } }, '2026-08-24', hcw2, CITIES),
  { reason: 'שבת', kind: 'shabbat', startIso: '2026-08-28', candle: '18:42', endIso: '2026-08-29', tzeis: '19:53', cityHe: 'תל אביב' });
eq('דוגמה 6 (ארגומנטים-לשקע)', gotArgs, ['2026-08-24', 10, { city: 'telaviv', timezone: 'Asia/Jerusalem' }, {}]);

// 3. אין חלונות ⇒ null
eq('דוגמה 3 (אין חלון)', nextClosure({ telephony: { city: 'telaviv' } }, '2026-08-24', () => [], CITIES), null);

// 4. בלי city ⇒ tenant.city='default' + נפילת-ירושלים
let gotArgs4 = null;
const r4 = nextClosure({ telephony: {} }, '2026-08-24', (...a) => { gotArgs4 = a; return [WIN]; }, CITIES);
eq('דוגמה 4 (tenant default)', gotArgs4[2], { city: 'default', timezone: 'Asia/Jerusalem' });
eq('דוגמה 4 (cityHe ירושלים)', r4.cityHe, 'ירושלים');

// 5. עיר לא-מוכרת ⇒ tenant שומר אותה, cityHe נופל לירושלים
let gotArgs5 = null;
const r5 = nextClosure({ telephony: { city: 'nowhere' } }, '2026-08-24', (...a) => { gotArgs5 = a; return [WIN]; }, CITIES);
eq('דוגמה 5 (tenant nowhere)', gotArgs5[2].city, 'nowhere');
eq('דוגמה 5 (cityHe ירושלים)', r5.cityHe, 'ירושלים');

if (f) process.exit(1);
console.log('✓ next-closure: 6 דוגמאות-חוזה — ירוק');
